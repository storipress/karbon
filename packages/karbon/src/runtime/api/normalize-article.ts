import { destr } from 'destr'
import truncate from 'lodash.truncate'
import type { Segment } from '../lib/split-article'
import type { ArticleDesk, ArticlePlan, ArticleTag } from '../types'
import { useArticleFilter } from '#imports'

export interface RawUserLike {
  id: string
  slug: string
  first_name: string
  last_name: string
  full_name: string
  socials: Record<string, string>
}

export interface RawArticleLike {
  id: string
  title: string
  blurb: string
  bio: string
  seo: string
  cover: string
  html: string
  plaintext: string
  plan: ArticlePlan
  authors: RawUserLike[]
  tags: ArticleTag[]
  desk: ArticleDesk
  published_at: string | number
}

export interface PaidContent {
  content: string
  key: string
}

export interface RawSEOItem {
  title: string
  description: string
}

export interface RawSEO {
  meta?: RawSEOItem
  og?: RawSEOItem
  ogImage?: string
}

export function normalizeArticle({
  title,
  blurb,
  bio,
  seo,
  cover,
  plan,
  plaintext,
  html,
  id,
  authors,
  desk,
  tags,
  published_at,
  ...rest
}: RawArticleLike) {
  const articleFilter = useArticleFilter()
  const rootDesk = desk.desk ? { desk: { ...desk.desk, id: String(desk.desk.id) } } : {}

  return {
    ...rest,
    id,
    plan,
    bio: articleFilter(bio),
    bioHTML: bio,
    // published_at could be unix timestamp
    published_at: typeof published_at === 'string' ? published_at : new Date(published_at * 1000).toISOString(),
    title: unwrapParagraph(title),
    blurb: unwrapParagraph(blurb),
    seo: destr<RawSEO>(seo),
    html,
    plaintext: truncate(plaintext, {
      // description length
      length: 120,
      // don't cut on word
      separator: /,? +/,
    }),
    cover: destr(cover),
    authors: authors?.map(({ socials, id, ...rest }) => ({
      ...rest,
      id: String(id),
      socials: destr(socials),
      name: rest.full_name,
    })),
    desk: {
      ...desk,
      ...rootDesk,
      id: String(desk.id),
    },
    tags: tags?.map(({ id, ...rest }) => ({
      ...rest,
      id: String(id),
    })),
  }
}

export type _NormalizeArticle = ReturnType<typeof normalizeArticle>

export interface NormalizeArticle extends ReturnType<typeof normalizeArticle> {
  paidContent?: PaidContent
  segments: Segment[]
}

export function unwrapParagraph(input: string): string {
  if (!input) {
    return ''
  }

  return input.replace(/^<p>/, '').replace(/<\/p>$/, '')
}
