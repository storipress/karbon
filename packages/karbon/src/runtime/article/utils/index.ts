import type { UseAsyncStateReturn } from '@vueuse/core'
import { createInjectionState } from '@vueuse/core'
import invariant from 'tiny-invariant'
import { _useUniversalStoripressPayload } from '../../composables/storipress-payload'
import type { BaseMeta, UseArticleReturn } from '../../types'
import type { Segment } from '../../lib/split-article'
import {
  loadStoripressPayload,
  ref,
  shallowReadonly,
  useAsyncData,
  useFillArticles,
  useResourceResolver,
} from '#imports'

export { useSite } from '../../composables/site'

export function useGetArticle(id: string): UseAsyncStateReturn<any, [], true> {
  return _useUniversalStoripressPayload('posts', id)
}

const advertisingHandler = ref((segments: Segment[]) => segments)

export function useAdvertisingHandler(handler: (segments: Segment[]) => Segment[]) {
  advertisingHandler.value = handler
}

const [useProvideArticle, useOptionalArticle] = createInjectionState<[UseArticleReturn], UseArticleReturn>((article) =>
  shallowReadonly({
    ...article,
    segments: advertisingHandler.value(article.segments ?? []),
  })
)

export { useProvideArticle, useOptionalArticle }

export function useArticle() {
  const res = useOptionalArticle()
  invariant(res, 'must wrap in useProviderArticle')
  return res
}

type Article = UseArticleReturn

type ArticleWithURL = Article & { url: string }

export interface RecommendArticleOptions {
  count: number
  cacheKey?: string
}

export function getRelevancesArticles(article: Article, options: RecommendArticleOptions) {
  const { count } = options
  const { relevances } = article
  const relevancesArticlesPromise = (relevances ?? []).slice(0, count).map(async ({ id }): Promise<ArticleWithURL> => {
    const article = await loadStoripressPayload<Article>('posts', id)
    return enhanceArticleWithURL(article)
  })
  return Promise.all(relevancesArticlesPromise)
}

export function useRecommendArticle(article: Article, options: RecommendArticleOptions) {
  const { count } = options
  const baseCacheKey = `recommend-${article.id}-${count}`
  const { articles: fillArticles } = useFillArticles(count, [{ key: 'id', value: article.id }], {
    cacheKey: options.cacheKey || `${baseCacheKey}-fill`,
  })
  const { data } = useAsyncData(
    baseCacheKey,
    async () => {
      const relevancesArticles = await getRelevancesArticles(article, options)
      const alreadyRecommend = new Set<string>(relevancesArticles.map(({ id }) => id))
      const extraArticles = fillArticles.value
        .filter(({ id }) => !alreadyRecommend.has(id))
        .map((article) => enhanceArticleWithURL(article))

      const recommendArticles = [...relevancesArticles, ...extraArticles]

      return recommendArticles.slice(0, count)
    },
    {
      default: () => [],
      watch: [fillArticles],
    }
  )

  return data
}

function enhanceArticleWithURL(article: Article) {
  const { _resolveFromMetaSync } = useResourceResolver()
  const res = _resolveFromMetaSync('article', article as unknown as BaseMeta)
  invariant(res, 'impossible to resolve fail')
  return {
    ...article,
    url: res.url,
  }
}
