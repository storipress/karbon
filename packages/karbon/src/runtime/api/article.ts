import { Buffer } from 'node:buffer'
import { gql } from '@apollo/client/core/index.js'
import { encrypt } from 'micro-aes-gcm'
import { destr } from 'destr'

// This file contains global crypto polyfill
import { CompactEncrypt } from '@storipress/jose-browser'
import type { SearchResponse } from 'typesense/lib/Typesense/Documents'
import { useStoripressClient } from '../composables/storipress-client'
import type { TypesenseFilter } from '../composables/typesense-client'
import { PER_PAGE, getSearchQuery, useTypesenseClient } from '../composables/typesense-client'
import { splitPaidContent } from '../lib/split-paid-content'
import type { NormalSegment } from '../lib/split-article'
import { splitArticle } from '../lib/split-article'
import { _karbonClientHooks, getStoripressConfig } from '../composables/storipress-base-client'
import { verboseInvariant } from '../utils/verbose-invariant'
import type { PaidContent, RawArticleLike, TypesenseArticleLike, _NormalizeArticle } from './normalize-article'
import { ArticleSchema } from './schema/typesense-article'
import { QueryArticleSchema } from './schema/query-article'
import { normalizeArticle } from './normalize-article'

export type { NormalizeArticle, PaidContent } from './normalize-article'

const GetArticle = gql`
  query GetArticle($id: ID!) {
    article(id: $id) {
      id
      blurb
      published_at
      updated_at
      desk {
        id
        name
        slug
        layout {
          id
          name
        }
        desk {
          id
          name
          slug
          layout {
            id
            name
          }
        }
      }
      slug
      title
      featured
      cover
      seo
      html
      plaintext
      layout {
        id
        name
      }
      tags {
        id
        slug
        name
      }
      authors {
        id
        bio
        slug
        socials
        avatar
        email
        location
        first_name
        last_name
        full_name
      }
      shadow_authors
      plan
      metafields {
        id
        key
        type
        values {
          __typename
          ... on CustomFieldTextValue {
            id
            value
          }
          ... on CustomFieldNumberValue {
            id
            numberValue: value
          }
          ... on CustomFieldColorValue {
            id
            value
          }
          ... on CustomFieldUrlValue {
            id
            value
          }
          ... on CustomFieldBooleanValue {
            id
            booleanValue: value
          }
          ... on CustomFieldRichTextValue {
            id
            jsonValue: value
          }
          ... on CustomFieldFileValue {
            id
            fileValue: value {
              key
              url
              size
              mime_type
            }
          }
          ... on CustomFieldDateValue {
            id
            dateValue: value
          }
          ... on CustomFieldJsonValue {
            id
            jsonValue: value
          }
          ... on CustomFieldReferenceValue {
            id
            referenceValue: value {
              ... on Article {
                __typename
                id
                articleId: id
                title
              }
              ... on Desk {
                __typename
                id
                deskId: id
                name
              }
              ... on Tag {
                __typename
                id
                tagId: id
                name
              }
              ... on User {
                __typename
                id
                userId: id
                full_name
              }
            }
          }
          ... on CustomFieldSelectValue {
            id
            selectValue: value
          }
        }
        group {
          id
          key
          type
        }
      }
      relevances {
        id
        title
      }
      content_blocks {
        id
        key
        type
        values {
          __typename
          ... on CustomFieldTextValue {
            id
            value
          }
          ... on CustomFieldNumberValue {
            id
            numberValue: value
          }
          ... on CustomFieldColorValue {
            id
            value
          }
          ... on CustomFieldUrlValue {
            id
            value
          }
          ... on CustomFieldBooleanValue {
            id
            booleanValue: value
          }
          ... on CustomFieldRichTextValue {
            id
            jsonValue: value
          }
          ... on CustomFieldFileValue {
            id
            fileValue: value {
              key
              url
              size
              mime_type
            }
          }
          ... on CustomFieldDateValue {
            id
            dateValue: value
          }
          ... on CustomFieldJsonValue {
            id
            jsonValue: value
          }
          ... on CustomFieldReferenceValue {
            id
            referenceValue: value {
              ... on Article {
                __typename
                id
                articleId: id
                title
              }
              ... on Desk {
                __typename
                id
                deskId: id
                name
              }
              ... on Tag {
                __typename
                id
                tagId: id
                name
              }
              ... on User {
                __typename
                id
                userId: id
                full_name
              }
            }
          }
          ... on CustomFieldSelectValue {
            id
            selectValue: value
          }
        }
        group {
          id
          key
          type
        }
      }
    }
  }
`

export async function listArticles(filter?: TypesenseFilter) {
  const storipress = getStoripressConfig()
  const typesenseClient = useTypesenseClient()
  const documents = typesenseClient.collections('articles').documents()

  const groupId = crypto.randomUUID()
  const site = storipress.clientId
  const name = 'listArticles'
  const articles = []
  const groupStartTime = Date.now()
  let hasMore = true
  let page = 1
  while (hasMore) {
    const id = crypto.randomUUID()
    const query = getSearchQuery(page, filter)
    const isFirstRequest = page === 1
    const ctx = { name, id, groupId, query, site, isFirstRequest, groupStartTime, requestTime: Date.now() }
    _karbonClientHooks.callHookParallel('karbon:searchRequest', ctx)

    // `destr` is workaround for fetch adapter not automatically parse response
    const rawSearchResult = await documents.search(query, {}).catch((error: Error) => {
      _karbonClientHooks.callHookParallel('karbon:searchResponse', {
        ...ctx,
        type: 'error',
        responseTime: Date.now(),
        hasMore,
        error,
      })
      throw error
    })
    const searchResult = destr<SearchResponse<TypesenseArticleLike>>(rawSearchResult)

    const failedDocument: TypesenseArticleLike[] = []
    const currentPageArticles =
      searchResult?.hits?.map(({ document }) => {
        if (process.env.NODE_ENV === 'development') {
          const { success } = ArticleSchema.safeParse(document)
          if (!success) failedDocument.push(document)
        }
        const article = normalizeArticle(document as TypesenseArticleLike)
        return article
      }) ?? []
    if (failedDocument.length > 0) {
      console.error(new Error(`Invalid article: ${JSON.stringify(failedDocument)}`))
    }
    articles.push(...currentPageArticles)

    hasMore = searchResult.found > searchResult.page * PER_PAGE
    page = searchResult.page + 1

    _karbonClientHooks.callHookParallel('karbon:searchResponse', {
      ...ctx,
      type: 'complete',
      hasMore,
      responseTime: Date.now(),
    })
  }
  return articles
}

export async function getArticle(id: string) {
  const client = useStoripressClient()
  const { data } = await client.query<{ article: RawArticleLike }>({ query: GetArticle, variables: { id } })
  if (!data || !data.article) {
    return null
  }

  if (process.env.NODE_ENV === 'development') {
    const { success } = QueryArticleSchema.safeParse(document)
    if (!success) {
      console.error(new Error(`Invalid article: ${JSON.stringify(document)}`))
    }
  }
  const res = await encryptArticle(normalizeArticle(data.article))
  return res
}

async function encryptArticle({ plan, html, id, ...rest }: _NormalizeArticle) {
  let freeHTML = html
  let paidContent: PaidContent | undefined

  let segments = splitArticle(html)

  if (plan !== 'free') {
    const storipress = getStoripressConfig()
    verboseInvariant(storipress.encryptKey, 'No encrypt key')
    const previewParagraph = storipress.previewParagraph ?? 3
    const [preview, paid] = splitPaidContent(html, storipress.previewParagraph ?? 3)
    freeHTML = preview

    const cryptoKey = await crypto.subtle.generateKey({ name: 'AES-GCM', length: 256 }, true, ['encrypt'])
    const key = await crypto.subtle.exportKey('raw', cryptoKey)
    const content = await encrypt(new Uint8Array(key), paid)
    const compactEncrypter = new CompactEncrypt(
      Buffer.from(JSON.stringify({ id, plan, key: Buffer.from(key).toString('base64') })),
    ).setProtectedHeader({ enc: 'A256GCM', alg: 'dir' })
    const encryptedKey = await compactEncrypter.encrypt(Buffer.from(storipress.encryptKey, 'base64'))
    paidContent = {
      key: encryptedKey,
      content: Buffer.from(content).toString('base64'),
    }

    segments = await Promise.all(
      segments.map(async (segment, index, source) => {
        const html = (segment as NormalSegment).html
        const noEncrypt = html === undefined || (index < previewParagraph && source.length > previewParagraph)
        if (noEncrypt) return segment

        const content = await encrypt(new Uint8Array(key), html)
        return {
          id: 'paid',
          type: segment.type,
          paidContent: Buffer.from(content).toString('base64'),
        }
      }),
    )
  }

  return {
    ...rest,
    id,
    plan,
    html: freeHTML,
    paidContent,
    segments,
  }
}
