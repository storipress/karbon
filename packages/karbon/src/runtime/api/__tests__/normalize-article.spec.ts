import { expect, it, vi } from 'vitest'
import { normalizeArticle } from '../normalize-article'
import { mockQueryArticle, mockTypesenseArticle } from './article.mock'

vi.mock('#imports', async () => await vi.importActual('../../../normalize-helper'))

it('mockTypesenseArticle', () => {
  expect(normalizeArticle(mockTypesenseArticle)).toMatchSnapshot()
  expect(normalizeArticle(mockQueryArticle)).toMatchSnapshot()
})
