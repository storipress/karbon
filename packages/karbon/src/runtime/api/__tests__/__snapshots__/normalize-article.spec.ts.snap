// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`mockTypesenseArticle 1`] = `
{
  "authors": [
    {
      "avatar": "https://assets.stori.press/media/images/f3354bcf-a572-4fc0-8e3d-d19cbcfbc318.png",
      "bio": "<p>Sid Test</p>",
      "full_name": "Sidd Chang",
      "id": "242",
      "location": "T",
      "name": "Sidd Chang",
      "slug": "sidd-chang",
      "socials": [],
    },
  ],
  "bio": "Sid Test",
  "bioHTML": "<p>Sid Test</p>",
  "blurb": "This is a mock article",
  "cover": {
    "alt": null,
    "caption": "<p>Photo by <a href=\\"https://unsplash.com/@kz29?utm_source=Storipress&utm_medium=referral&utm_campaign=api-credit?utm_source=storipress&utm_medium=referral&utm_campaign=api-credit\\">Keith Zhang</a> / <a href=\\"https://unsplash.com/?utm_source=storipress&utm_medium=referral&utm_campaign=api-credit\\">Unsplash</a></p>",
    "crop": {
      "height": 0,
      "key": "",
      "left": 50,
      "realHeight": 3024,
      "realWidth": 4032,
      "top": 50,
      "width": 0,
      "zoom": 1,
    },
    "url": "https://images.unsplash.com/photo-1704461539031-e7c2145e5fc4?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=M3wyMzc2NzF8MHwxfGFsbHw5fHx8fHx8Mnx8MTcwNDc1ODc2Nnw&ixlib=rb-4.0.3&q=80&w=1080",
  },
  "desk": {
    "id": "119",
    "name": "test",
    "slug": "test-32",
  },
  "featured": false,
  "html": "",
  "id": "619",
  "order": 400,
  "pathnames": [
    "/posts/mock-article",
    "/posts/mock-article",
    "/posts/mock-article",
    "/posts/new-mock-article",
    "/posts/mock-article",
  ],
  "plaintext": "",
  "plan": "free",
  "published_at": "2024-01-08T23:46:33.000Z",
  "seo": {
    "hasSlug": true,
    "meta": {
      "description": "Discover expert tips for boosting your website's visibility and driving traffic with our comprehensive SEO strategies guide. strategies guide.",
      "title": "SEO Title Tag: Essential Tips for Effective Article Writing Article Writing",
    },
    "og": {
      "description": "",
      "title": "",
    },
    "ogImage": "",
  },
  "slug": "mock-article-test",
  "tags": [
    {
      "id": "44",
      "name": "foo",
      "slug": "foo",
    },
  ],
  "title": "Mock Article",
  "updated_at": 1704758778,
}
`;

exports[`mockTypesenseArticle 2`] = `
{
  "authors": [
    {
      "avatar": "https://assets.stori.press/media/images/f3354bcf-a572-4fc0-8e3d-d19cbcfbc318.png",
      "bio": "<p>Sid Test</p>",
      "email": "sid+0705@storipress.com",
      "first_name": "Sidd",
      "full_name": "Sidd Chang",
      "id": "242",
      "last_name": "Chang",
      "location": "T",
      "name": "Sidd Chang",
      "slug": "sidd-chang",
      "socials": [],
    },
  ],
  "bio": "Sid Test",
  "bioHTML": "<p>Sid Test</p>",
  "blurb": "This is a mock article",
  "content_blocks": [],
  "cover": {
    "alt": null,
    "caption": "<p>Photo by <a href=\\"https://unsplash.com/@kz29?utm_source=Storipress&utm_medium=referral&utm_campaign=api-credit?utm_source=storipress&utm_medium=referral&utm_campaign=api-credit\\">Keith Zhang</a> / <a href=\\"https://unsplash.com/?utm_source=storipress&utm_medium=referral&utm_campaign=api-credit\\">Unsplash</a></p>",
    "crop": {
      "height": 0,
      "key": "",
      "left": 50,
      "realHeight": 3024,
      "realWidth": 4032,
      "top": 50,
      "width": 0,
      "zoom": 1,
    },
    "url": "https://images.unsplash.com/photo-1704461539031-e7c2145e5fc4?crop=entropy&cs=tinysrgb&fit=max&fm=jpg&ixid=M3wyMzc2NzF8MHwxfGFsbHw5fHx8fHx8Mnx8MTcwNDc1ODc2Nnw&ixlib=rb-4.0.3&q=80&w=1080",
  },
  "desk": {
    "desk": null,
    "id": "119",
    "layout": null,
    "name": "test",
    "slug": "test-32",
  },
  "featured": false,
  "html": "<p>Mock Article Content</p>",
  "id": "619",
  "layout": null,
  "metafields": [
    {
      "group": {
        "id": "70",
        "key": "__layoutmeta",
        "type": "articleMetafield",
      },
      "id": "76",
      "key": "layoutid",
      "type": "text",
      "values": [],
    },
    {
      "group": {
        "id": "81",
        "key": "test",
        "type": "articleMetafield",
      },
      "id": "100",
      "key": "test",
      "type": "text",
      "values": [],
    },
    {
      "group": {
        "id": "82",
        "key": "webflow",
        "type": "articleMetafield",
      },
      "id": "104",
      "key": "color",
      "type": "color",
      "values": [],
    },
    {
      "group": {
        "id": "82",
        "key": "webflow",
        "type": "articleMetafield",
      },
      "id": "103",
      "key": "thumbnail-image",
      "type": "file",
      "values": [],
    },
  ],
  "plaintext": "Mock Article Content",
  "plan": "free",
  "published_at": "2024-01-08T23:46:33+00:00",
  "relevances": [
    {
      "id": "21",
      "title": "When an editor thinks an article is",
    },
    {
      "id": "20",
      "title": "When a writer wants an editor to review their work, it goes here",
    },
    {
      "id": "24",
      "title": "Chris' First Article Chris' First ArticleChris' First ArticleChris' First ArticleChris' First ArticleChris' First Article",
    },
    {
      "id": "43",
      "title": "<p>&lt;c&gt;12345&lt;/c&gt;</p>",
    },
    {
      "id": "596",
      "title": "<p>New Article 2</p>",
    },
    {
      "id": "597",
      "title": "<p>New article 3</p>",
    },
    {
      "id": "598",
      "title": "<p>New article 4</p>",
    },
  ],
  "seo": {
    "hasSlug": true,
    "meta": {
      "description": "Discover expert tips for boosting your website's visibility and driving traffic with our comprehensive SEO strategies guide. strategies guide.",
      "title": "SEO Title Tag: Essential Tips for Effective Article Writing Article Writing",
    },
    "og": {
      "description": "",
      "title": "",
    },
    "ogImage": "",
  },
  "shadow_authors": null,
  "slug": "mock-article-test",
  "tags": [
    {
      "id": "44",
      "name": "foo",
      "slug": "foo",
    },
  ],
  "title": "Mock Article",
  "updated_at": "2024-01-09T00:32:10+00:00",
}
`;
